File: apps/desktop/electron/db/mysql.ts
import mysql from 'mysql2/promise'
import { DatabaseConnection, ConnectionConfig, StorageAnalysis, QueryResult, SchemaInfo } from './index'
import { writeFileSync, readFileSync } from 'fs'

export class MySQLConnection extends DatabaseConnection {
  private connection?: mysql.Connection

  async connect(): Promise<void> {
    this.connection = await mysql.createConnection({
      host: this.config.host,
      port: this.config.port || 3306,
      database: this.config.database,
      user: this.config.username,
      password: this.config.password,
      connectionString: this.config.connectionString,
    })

    this.connected = true
  }

  async disconnect(): Promise<void> {
    if (this.connection) {
      await this.connection.end()
      this.connection = undefined
      this.connected = false
    }
  }

  async testConnection(): Promise<void> {
    if (!this.connection) {
      await this.connect()
    }
    
    const [rows] = await this.connection!.execute('SELECT 1 as test')
    if (!rows || (rows as any[]).length === 0) {
      throw new Error('Connection test failed')
    }
  }

  async analyzeStorage(): Promise<StorageAnalysis> {
    if (!this.connection) {
      await this.connect()
    }

    const connection = this.connection!

    // Get database size
    const [dbSizeResult] = await connection.execute(`
      SELECT 
        ROUND(SUM(data_length + index_length) / 1024 / 1024, 2) AS size_mb,
        SUM(data_length + index_length) AS size_bytes
      FROM information_schema.tables 
      WHERE table_schema = ?
    `, [this.config.database])

    const totalSize = (dbSizeResult as any[])[0]?.size_bytes || 0

    // Get table information
    const [tablesResult] = await connection.execute(`
      SELECT 
        table_name,
        ROUND((data_length + index_length) / 1024 / 1024, 2) AS size_mb,
        (data_length + index_length) AS size_bytes,
        ROUND(data_length / 1024 / 1024, 2) AS table_size_mb,
        data_length AS table_size_bytes,
        ROUND(index_length / 1024 / 1024, 2) AS index_size_mb,
        index_length AS index_size_bytes,
        table_rows AS row_count
      FROM information_schema.tables 
      WHERE table_schema = ?
      ORDER BY (data_length + index_length) DESC
    `, [this.config.database])

    const tables = (tablesResult as any[]).map(row => ({
      name: row.table_name,
      size: row.size_bytes,
      rowCount: row.row_count || 0,
      indexSize: row.index_size_bytes,
      bloat: 0 // Placeholder for bloat calculation
    }))

    // Get index information
    const [indexesResult] = await connection.execute(`
      SELECT 
        index_name,
        table_name,
        ROUND(stat_value * @@innodb_page_size / 1024 / 1024, 2) AS size_mb,
        stat_value * @@innodb_page_size AS size_bytes
      FROM information_schema.innodb_buffer_stats_by_table 
      WHERE object_schema = ?
      ORDER BY stat_value * @@innodb_page_size DESC
    `, [this.config.database])

    const indexes = (indexesResult as any[]).map(row => ({
      name: row.index_name,
      tableName: row.table_name,
      size: row.size_bytes,
      bloat: 0 // Placeholder for bloat calculation
    }))

    const largestTable = tables.length > 0 ? tables[0] : { name: 'N/A', size: 0, rowCount: 0 }

    return {
      totalSize,
      tableCount: tables.length,
      indexCount: indexes.length,
      largestTable,
      tables,
      indexes,
      lastAnalyzed: new Date().toISOString()
    }
  }

  async executeQuery(query: string): Promise<QueryResult> {
    if (!this.connection) {
      await this.connect()
    }

    const startTime = Date.now()
    const [rows, fields] = await this.connection!.execute(query)
    const executionTime = Date.now() - startTime

    // Get explain plan if it's a SELECT query
    let explainPlan
    if (query.trim().toLowerCase().startsWith('select')) {
      try {
        const [explainResult] = await this.connection!.execute(`EXPLAIN ${query}`)
        explainPlan = explainResult
      } catch (error) {
        // Ignore explain errors
      }
    }

    return {
      columns: fields?.map(field => field.name) || [],
      rows: rows as any[],
      rowCount: (rows as any[]).length,
      executionTime,
      explainPlan
    }
  }

  async getSchema(): Promise<SchemaInfo> {
    if (!this.connection) {
      await this.connect()
    }

    const connection = this.connection!

    // Get tables
    const [tablesResult] = await connection.execute(`
      SELECT table_name 
      FROM information_schema.tables 
      WHERE table_schema = ?
      ORDER BY table_name
    `, [this.config.database])

    const tables = []

    for (const tableRow of tablesResult as any[]) {
      const tableName = tableRow.table_name

      // Get columns
      const [columnsResult] = await connection.execute(`
        SELECT 
          column_name,
          data_type,
          is_nullable,
          column_default
        FROM information_schema.columns 
        WHERE table_schema = ? AND table_name = ?
        ORDER BY ordinal_position
      `, [this.config.database, tableName])

      const columns = (columnsResult as any[]).map(col => ({
        name: col.column_name,
        type: col.data_type,
        nullable: col.is_nullable === 'YES',
        defaultValue: col.column_default
      }))

      // Get indexes
      const [indexesResult] = await connection.execute(`
        SELECT 
          index_name,
          column_name,
          non_unique
        FROM information_schema.statistics 
        WHERE table_schema = ? AND table_name = ?
        ORDER BY index_name, seq_in_index
      `, [this.config.database, tableName])

      const indexMap = new Map()
      for (const idxRow of indexesResult as any[]) {
        const indexName = idxRow.index_name
        if (!indexMap.has(indexName)) {
          indexMap.set(indexName, {
            name: indexName,
            columns: [],
            unique: idxRow.non_unique === 0
          })
        }
        indexMap.get(indexName).columns.push(idxRow.column_name)
      }

      const indexes = Array.from(indexMap.values())

      tables.push({
        name: tableName,
        columns,
        indexes
      })
    }

    return {
      schemas: [{
        name: this.config.database || 'default',
        tables
      }]
    }
  }

  async createBackup(backupPath: string): Promise<{ path: string; size: number }> {
    if (!this.connection) {
      await this.connect()
    }

    // For MySQL, we'll use mysqldump via child_process
    const { spawn } = require('child_process')

    return new Promise((resolve, reject) => {
      const mysqldump = spawn('mysqldump', [
        '--host', this.config.host || 'localhost',
        '--port', (this.config.port || 3306).toString(),
        '--user', this.config.username || '',
        '--password=' + (this.config.password || ''),
        '--single-transaction',
        '--routines',
        '--triggers',
        this.config.database || ''
      ])

      const fs = require('fs')
      const writeStream = fs.createWriteStream(backupPath)

      mysqldump.stdout.pipe(writeStream)

      mysqldump.on('close', async (code: number) => {
        if (code === 0) {
          try {
            const stats = await require('fs').promises.stat(backupPath)
            resolve({ path: backupPath, size: stats.size })
          } catch (error) {
            reject(new Error('Failed to get backup file size'))
          }
        } else {
          reject(new Error(`mysqldump failed with code ${code}`))
        }
      })

      mysqldump.on('error', (error: Error) => {
        reject(new Error(`mysqldump error: ${error.message}`))
      })
    })
  }

  async restoreBackup(backupPath: string): Promise<void> {
    if (!this.connection) {
      await this.connect()
    }

    const { spawn } = require('child_process')

    return new Promise((resolve, reject) => {
      const mysql = spawn('mysql', [
        '--host', this.config.host || 'localhost',
        '--port', (this.config.port || 3306).toString(),
        '--user', this.config.username || '',
        '--password=' + (this.config.password || ''),
        this.config.database || ''
      ])

      const fs = require('fs')
      const readStream = fs.createReadStream(backupPath)
      readStream.pipe(mysql.stdin)

      mysql.on('close', (code: number) => {
        if (code === 0) {
          resolve()
        } else {
          reject(new Error(`mysql restore failed with code ${code}`))
        }
      })

      mysql.on('error', (error: Error) => {
        reject(new Error(`mysql restore error: ${error.message}`))
      })
    })
  }
}

File: db_storage_manager/db/mysql.py
"""
MySQL/MariaDB database connection
"""

import asyncio
from pathlib import Path
from typing import Any, Dict, List
import pymysql
from pymysql.cursors import DictCursor

from .base import (
    DatabaseConnection,
    ConnectionConfig,
    StorageAnalysis,
    QueryResult,
    SchemaInfo,
)


class MySQLConnection(DatabaseConnection):
    """MySQL/MariaDB database connection"""

    def __init__(self, config: ConnectionConfig):
        super().__init__(config)
        self.connection = None

    async def connect(self) -> None:
        """Connect to MySQL database"""
        if self.config.connection_string:
            self.connection = pymysql.connect(
                host=self.config.host or "localhost",
                port=self.config.port or 3306,
                database=self.config.database,
                user=self.config.username,
                password=self.config.password,
            )
        else:
            self.connection = pymysql.connect(
                host=self.config.host or "localhost",
                port=self.config.port or 3306,
                database=self.config.database,
                user=self.config.username,
                password=self.config.password,
            )
        self.connected = True

    async def disconnect(self) -> None:
        """Disconnect from MySQL database"""
        if self.connection:
            self.connection.close()
            self.connection = None
        self.connected = False

    async def analyze_storage(self) -> StorageAnalysis:
        """Analyze MySQL storage"""
        if not self.connection:
            await self.connect()

        cursor = self.connection.cursor(DictCursor)

        # Get table sizes
        cursor.execute("""
            SELECT 
                table_schema,
                table_name,
                data_length + index_length as total_size,
                data_length as table_size,
                index_length as index_size,
                table_rows as row_count
            FROM information_schema.tables
            WHERE table_schema = %s
            ORDER BY total_size DESC
        """, (self.config.database,))

        tables = []
        total_size = 0
        for row in cursor.fetchall():
            table_name = f"{row['table_schema']}.{row['table_name']}"
            table_size = int(row["table_size"] or 0)
            index_size = int(row["index_size"] or 0)
            row_count = int(row["row_count"] or 0)

            tables.append({
                "name": table_name,
                "size": table_size,
                "rowCount": row_count,
                "indexSize": index_size,
                "bloat": 0.0,
            })
            total_size += int(row["total_size"] or 0)

        # Get indexes
        cursor.execute("""
            SELECT 
                index_name,
                table_name,
                SUM(index_length) as size
            FROM information_schema.statistics
            WHERE table_schema = %s
            GROUP BY index_name, table_name
        """, (self.config.database,))

        indexes = []
        for row in cursor.fetchall():
            indexes.append({
                "name": row["index_name"],
                "tableName": row["table_name"],
                "size": int(row["size"] or 0),
                "bloat": 0.0,
            })

        largest_table = tables[0] if tables else {"name": "N/A", "size": 0, "rowCount": 0}

        return {
            "totalSize": total_size,
            "tableCount": len(tables),
            "indexCount": len(indexes),
            "largestTable": largest_table,
            "tables": tables,
            "indexes": indexes,
            "lastAnalyzed": __import__("datetime").datetime.now().isoformat(),
        }

    async def execute_query(self, query: str, safe_mode: bool = True) -> QueryResult:
        """Execute a MySQL query"""
        if not self.connection:
            await self.connect()

        if safe_mode and not query.strip().upper().startswith("SELECT"):
            raise ValueError("Only SELECT queries are allowed in safe mode")

        cursor = self.connection.cursor(DictCursor)
        import time
        start_time = time.time()

        try:
            cursor.execute(query)
            if cursor.description:
                columns = [desc[0] for desc in cursor.description]
                rows = [dict(row) for row in cursor.fetchall()]
            else:
                columns = ["affected_rows"]
                rows = [{"affected_rows": cursor.rowcount}]

            execution_time = int((time.time() - start_time) * 1000)

            # Get explain plan for SELECT queries
            explain_plan = None
            if query.strip().upper().startswith("SELECT"):
                try:
                    cursor.execute(f"EXPLAIN {query}")
                    explain_plan = [dict(row) for row in cursor.fetchall()]
                except Exception:
                    pass

            return {
                "columns": columns,
                "rows": rows,
                "rowCount": len(rows),
                "executionTime": execution_time,
                "explainPlan": explain_plan,
            }
        finally:
            cursor.close()

    async def get_schema(self) -> SchemaInfo:
        """Get MySQL schema"""
        if not self.connection:
            await self.connect()

        cursor = self.connection.cursor(DictCursor)

        # Get tables
        cursor.execute("""
            SELECT table_name 
            FROM information_schema.tables 
            WHERE table_schema = %s
        """, (self.config.database,))

        tables = []
        for table_row in cursor.fetchall():
            table_name = table_row["table_name"]

            # Get columns
            cursor.execute("""
                SELECT 
                    column_name,
                    data_type,
                    is_nullable,
                    column_default
                FROM information_schema.columns
                WHERE table_schema = %s AND table_name = %s
                ORDER BY ordinal_position
            """, (self.config.database, table_name))

            columns = []
            for col in cursor.fetchall():
                columns.append({
                    "name": col["column_name"],
                    "type": col["data_type"],
                    "nullable": col["is_nullable"] == "YES",
                    "defaultValue": col["column_default"],
                })

            # Get indexes
            cursor.execute("""
                SELECT 
                    index_name,
                    non_unique,
                    column_name
                FROM information_schema.statistics
                WHERE table_schema = %s AND table_name = %s
            """, (self.config.database, table_name))

            indexes_dict = {}
            for idx in cursor.fetchall():
                idx_name = idx["index_name"]
                if idx_name not in indexes_dict:
                    indexes_dict[idx_name] = {
                        "name": idx_name,
                        "columns": [],
                        "unique": idx["non_unique"] == 0,
                    }
                indexes_dict[idx_name]["columns"].append(idx["column_name"])

            indexes = list(indexes_dict.values())

            tables.append({
                "name": table_name,
                "columns": columns,
                "indexes": indexes,
            })

        return {
            "schemas": [{
                "name": self.config.database or "default",
                "tables": tables,
            }]
        }

    async def create_backup(self, backup_path: str) -> Dict[str, Any]:
        """Create MySQL backup using mysqldump"""
        if not self.connection:
            await self.connect()

        backup_file = Path(backup_path)
        backup_file.parent.mkdir(parents=True, exist_ok=True)

        # Build mysqldump command
        cmd = [
            "mysqldump",
            f"--host={self.config.host or 'localhost'}",
            f"--port={self.config.port or 3306}",
            f"--user={self.config.username}",
            f"--password={self.config.password}",
            self.config.database,
        ]

        # Run mysqldump
        process = await asyncio.create_subprocess_exec(
            *cmd,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )

        stdout, stderr = await process.communicate()

        if process.returncode != 0:
            raise RuntimeError(f"mysqldump failed: {stderr.decode()}")

        # Write output to file
        backup_file.write_bytes(stdout)
        return str(backup_file)

    async def restore_backup(self, backup_path: str) -> None:
        """Restore MySQL backup"""
        if not self.connection:
            await self.connect()

        backup_file = Path(backup_path)
        if not backup_file.exists():
            raise FileNotFoundError(f"Backup file not found: {backup_path}")

        # Read backup file
        backup_data = backup_file.read_bytes()

        # Build mysql command
        cmd = [
            "mysql",
            f"--host={self.config.host or 'localhost'}",
            f"--port={self.config.port or 3306}",
            f"--user={self.config.username}",
            f"--password={self.config.password}",
            self.config.database,
        ]

        # Run mysql restore
        process = await asyncio.create_subprocess_exec(
            *cmd,
            stdin=asyncio.subprocess.PIPE,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )

        stdout, stderr = await process.communicate(input=backup_data)

        if process.returncode != 0:
            raise RuntimeError(f"mysql restore failed: {stderr.decode()}")

